### PR TITLE
Add the response_format parameter to completion of Chatgpt

### PR DESCRIPTION
**Pull request by thelyad**
The completions function takes one more argument: ResponseFormat that can be either set of "text" or "json_object". 
We are guaranteed to obtain a Json object from ChatGPT is it is set to "json_object"

### CODE REVIEW

**Review by dan-nectry — APPROVED** [review]

File: src/ur/chatgpt.urs [hunk 1]
```diff
@@ -23,7 +23,8 @@ functor Make(M : sig
     structure Chat : sig
         val completions : {Model : model,
                            Messages : list {Role : role,
-                                            Content : string}}
+                                            Content : string},
+                            ResponseFormat: {Type: string}}
```
Comment (dan-nectry) [inline src/ur/chatgpt.urs]: Hmm, as I look at this, it feels a bit off.  Why not just `ResponseFormat: string`?

Also, my general feeling is that fields like this that should really only be one of a handful of values should be enumerations (e.g., `datatype ResponseFormat = AsJson | RegularText` or something).  Is there a reason you went with the more general `string`?

Lastly, I know these files can be super light on documentation, but all the more reason that when we make changes to them, we should add that documentation.  I realize looking at this now that I would have no idea what to pass as the value of `ResponseFormat` here.

**Comment by dan-nectry** [comment]
> ResponseFormat that can be either set of "text" or "json_object".

Yeah, this line needs to be in the code (or `datatype ResponseFormat = Text | Json_Object`)

File: src/ur/chatgpt.urs (hunk 1 above)
Comment (thelyad) [reply src/ur/chatgpt.urs]: >  Why not just 'ResponseFormat: string'

So according to the OpenAI documentation, the value has to be a dict such as `{type: "text"}`.

>  ... fields like this that should really only be one of a handful of values should be enumerations (e.g., datatype ResponseFormat = AsJson | RegularText or something)

I will update the code and add more documentations shortly.

Comment (thelyad) [reply src/ur/chatgpt.urs]: I just updated the code and have made corresponding changes to the nectry repo as well.

File: src/ur/chatgpt.ur [hunk 2]
```diff
@@ -53,14 +53,37 @@ type model = string
 val read_model = _
 val show_model = _
 
+datatype response_format = AsJson | RegularText
+type response_format' = variant [AsJson = unit, RegularText = unit]
```
Comment (dan-nectry) [inline src/ur/chatgpt.ur]: I think this line is superfluous

**Review by dan-nectry — APPROVED** [review]

File: src/ur/chatgpt.urs (hunk 1 above)
Comment (dan-nectry) [reply src/ur/chatgpt.urs]: > So according to the OpenAI documentation, the value has to be a dict such as `{type: "text"}`

Aha, _this_ needs to be in the code.  That is, if someone comes and looks at this code later, they may ask the natural question "why is it this way", and there's nothing about the structure of the code that naturally answers that.  I would put in a doc line that says something like (feel free to reword):

> The OpenAI documentation (found here: <link>) dictates that response type is a field of a dict with key "type".  There are other values allowed in this dict, but we are only using "type" for now.

I'm a big fan of docs answering the question "why".

Comment (thelyad) [reply src/ur/chatgpt.urs]: Sounds good. I just fixed it in my most recent commit.

**Pull request merged**